client/DESIGN.md

```diff
@@ -0,0 +1,292 @@
+
+The cache entries themselves are stored in (sorted value maps).  The keys of the (sorted value maps) are derived from the read timestamps of each entry.  Since we want the entries sorted from most-recent to oldest read timestamps, the actual keys are longs with the value 
+
+key = (LONG\_MAX - read\_timestamp)
+
+For example, in Python, the outer "table:key" map can be implemented as a standard Python dictionary, and we just use string concatenation (a+b) to form the keys.  For the inner sorted value maps, we can use a SortedDict implementation based on [skiplists](https://github.com/sepeth/python-skiplist).
+
+* **Question (1):** We said Tuesday we just need a single-level map, but why don't we need at least the two levels I describe here?
+
+## Cache Public Method Descriptions
+
+### read
+    read(
+        table: string, 
+        key: string, 
+        read_timestamp=currentTime, 
+        max_age=None, 
+        no_cache=False): value (JSON)  
+
+The read method sends a read request to the DB for the given table and key, along with the specified cache parameters, if any.  It returns the value object if the DB provides a value; otherwise, it returns None.
+
+In the case read returns None (value not in database), we still update our cache with the value entry None for the specified key at the read\_timestamp.  The None value may represent an update to the key.
+
+Note that the max\_age, and no\_cache parameters can be specified in at least three separate locations: (1) in the Cache constructor, (2) in the Cache read method, and (3) in the Transaction read method.  The cache read method determines the most restrictive set of parameters based on these three inputs and sends this most restrictive parameter set to the DB.  With respect to the read\_timestamp parameter, the DB server's response to every query includes the read\_timestamp it utilized to fulfill the request; we use the read\_timestamp the server returns to us to update the cache.
+
+Here is the method for determining the most restrictive parameter set:
+
+* **read\_timestamp**: The most restritive read\_timestamp is the minimum of the given read\_timestamp values (i.e., min(cache, transaction) read\_timestamps)
+* **max\_age**: The most restrictive max\_age is the minimum of all given max\_age values
+* **no\_cache**: The most restrictive no\_cache value is the disjunction (boolean "or") of all given no\_cache values.  (If no\_cache == True, the cache must verify with the server by checking the value timestamp it has the most up-to-date value.)
+
+### write
+    write(
+        table: string, 
+        op_List: list): value (JSON) throws StaleException  
+
+The write method sends the op\_List to the DB server as a batch write.  
+
+If the server accepts the batch write, the write method also updates the cache to reflect the newly written values.
+
+* **Question (2)**: What is the expected format of the op\_List?  (What are the possible types of its elements?)
```

Comment:
Ah, great question. The OPS are 

- Create (table, key, value)
- Hold (table, key)
- Update (table, key, value)
- Delete (table, key)

Note, the table its own arg to write. The table is inside each ops. That way, a write can update more than one table. Hold means that the write does not change the value, but it does depend on the value having remained unchanged since it was read.